### DISCUSSION TITLE
omp.h file not found, and visual studio code can't found omp.h

### DISCUSSION THREAD

**Discussion by marianofernandezs** [Homebrew/discussions#5282] [Everyday usage]
### Output of `brew config`

```shell
#include <omp.h> ## omp.h file not found
#include <stdio.h>
#include <stdlib.h>
int main(int argc, char* argv[]){
int nthreads = 3;
int tid;
printf("Antes del fork\n");
omp_set_num_threads(nthreads);
#pragma omp parallel private(nthreads, tid)
{
tid = omp_get_thread_num();
printf("Ejecutando thread %d - ",tid);
if (tid >= 0 && tid < (omp_get_num_threads() / 2 )){
printf("Soy productor\n");
}
else{
printf("Soy consumidor\n");
}
if(tid == 0){
printf("SOY EL PRINCIPAL: num_threads %d \n",
omp_get_num_threads());
}
}
printf("Despues del join");
}
```

### Output of `brew doctor`

```shell
id: library 'omp' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)

 *  The terminal process "/bin/zsh '-l', '-c', '/usr/local/opt/llvm/bin/clang -Xpreprocessor -fopenmp -lomp /Users/mariano/Downloads/c/programa.c -o /Users/mariano/Downloads/c/programa'" terminated with exit code: 1
```

### Description of issue

hi, I installed llvm and libomp in my MacBook Pro 16  Late 2019, I have a problem with visual studio can't run and compile my code, and the terminal return this:

id: library 'omp' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)

 *  The terminal process "/bin/zsh '-l', '-c', '/usr/local/opt/llvm/bin/clang -Xpreprocessor -fopenmp -lomp /Users/mariano/Downloads/c/programa.c -o /Users/mariano/Downloads/c/programa'" terminated with exit code: 1

I need help, I tried to chatgpt help but didn't get the solution he gave me. I tried to configure task.json but it didn't work.
If you land here, I would appreciate it.

**Comment by ZhongRuoyu** [comment]
What's the output of `brew info llvm`?

**Reply by marianofernandezs** [reply]
lvm: stable 17.0.6 (bottled), HEAD [keg-only]
Next-gen compiler infrastructure
https://llvm.org/
/usr/local/Cellar/llvm/17.0.6_1 (7,207 files, 1.8GB)
  Poured from bottle using the formulae.brew.sh API on 2024-03-24 at 23:52:48
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/l/llvm.rb
License: Apache-2.0 with LLVM-exception
==> Dependencies
Build: cmake ✔, ninja ✔, swig ✔
Required: python@3.12 ✔, z3 ✔, zstd ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
To use the bundled libc++ please add the following LDFLAGS:
  LDFLAGS="-L/usr/local/opt/llvm/lib/c++ -Wl,-rpath,/usr/local/opt/llvm/lib/c++"

llvm is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.

If you need to have llvm first in your PATH, run:
  echo 'export PATH="/usr/local/opt/llvm/bin:$PATH"' >> ~/.zshrc

For compilers to find llvm you may need to set:
  export LDFLAGS="-L/usr/local/opt/llvm/lib"
  export CPPFLAGS="-I/usr/local/opt/llvm/include"
==> Analytics
install: 37,205 (30 days), 138,519 (90 days), 499,172 (365 days)
install-on-request: 15,214 (30 days), 60,132 (90 days), 311,780 (365 days)
build-error: 338 (30 days)

This is my output

**Comment by ZhongRuoyu** [comment]
[`-Xpreprocessor`](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-Xpreprocessor) means its subsequent flag is passed to the preprocessor only, so the compiler driver is not passing the OpenMP-related ldflags to the linker. Removing the `-Xpreprocessor` flag should work.

**Reply by marianofernandezs** [reply]
Thanks, i don't work
![Screenshot 2024-04-12 at 10 45 19 PM](https://github.com/Homebrew/discussions/assets/164780229/96b0ea49-701a-42eb-b7f0-3eb120970a90)
![Screenshot 2024-04-12 at 10 45 50 PM](https://github.com/Homebrew/discussions/assets/164780229/4f038aab-b98f-4709-90f7-c75f7dc6da4e)

**Comment by trinitronx** [comment]
Looks like `tasks.json` is correct for the [keg-only][1] include dirs for `llvm`.  So compile via VSCode's "Tasks -> `build`" should work.  If you just need `libomp`, then you have a choice of using `brew install libomp` for the standalone runtime lib, or `brew install llvm` if using the one from LLVM.  For the `libomp` version, you can try adding this to the `build` task:

```json
{
  "tasks": [
    {
      "args": [
        "-L/usr/local/opt/libomp/lib",
        "-I/usr/local/opt/libomp/include",
        "-lomp"
      ]
    }
  ]
}
```

Note that in VSCode, the IntelliSense warnings are coming from a different set of commands that VSCode runs in the background (usually `clangd`, `clang-tidy`, or `oclint` when using LLVM), which are _not taking into account any specified `args` in `tasks.json`_ because they use separate configuration files.  You will usually need to generate a [`compile_commands.json`][2] first for VSCode with `clangd`, and most other linting extensions that use this [compilation database][2] to find the same includes and give the right IntelliSense warnings & output in the "Problems" tab.  For use with `llvm` & `clangd`, I've had some luck using the [`intercept-build` tool from `scan-build`][3] which can be installed via `pip` as a Python module. This command can then be prefixed to your compile command in `tasks.json` (e.g. `intercept-build --append ...make command + args here...`).  Alternatively, or in addition to the `compile_commands.json`, you can try using a `.clangd` file to configure the global include args for `clangd` to find:


`.clangd`:

```
CompileFlags:
    CompilationDatabase: "."
    Add:
    - "-Iexample/include"
    - "-Imore-includes/include"
```

[1]: https://github.com/Homebrew/brew/blob/master/docs/How-to-Build-Software-Outside-Homebrew-with-Homebrew-keg-only-Dependencies.md
[2]: https://clangd.llvm.org/design/compile-commands#where-do-compile-commands-come-from
[3]: https://pypi.org/project/scan-build/
